### PR TITLE
convert Choice to functional component with flow types and an interactive story

### PR DESCRIPTION
**Pull request by handeyeco**
## Summary:
In order to make Perseus easier to maintain, this PR refactors Choice (and ChoiceNoneAbove since it's so interconnected) to:

- use functional components instead of class components
- use Flow instead of PropTypes
- use interactive stories to be able to simulate all use-cases

I also:
- went through and tried to prune unused props from Choice.
- refactored some logic out of Choice and into OptionStatus since it's only used in Choice

The major issue was ref forwarding. You can attach a ref directly onto a class component, but not a functional component. Thanks Kevin for helping me navigate that!

Issue: LP-12661 (Also LP-12659 since it was so interconnected)

## Test plan:
- user-facing style and behavior should stay the same
- stories, types, and tests should be more comprehensive

### CODE REVIEW

File: packages/perseus/src/widgets/radio/__stories__/choice-icon.stories.jsx [hunk 1]
```diff
@@ -219,22 +219,28 @@ export const AllPositions = (args: StoryArgs): React.Node => {
         <>
             <p>Library</p>
             {allLetters.map((_, i: number) => (
-                <ChoiceIcon {...defaultProps} pos={i} />
+                <ChoiceIcon {...defaultProps} pos={i} key={"choice" + i} />
```
Comment (handeyeco) [inline packages/perseus/src/widgets/radio/__stories__/choice-icon.stories.jsx:222]: I added these to keep the console clean of errors when looking at this component in Storybook.

File: packages/perseus/src/widgets/radio/__stories__/option-status.stories.jsx [hunk 2]
```diff
@@ -22,6 +22,8 @@ export const CheckedCorrectPreviouslyAnswered = (
             checked={true}
             correct={true}
             previouslyAnswered={true}
+            reviewMode={true}
```
Comment (handeyeco) [inline packages/perseus/src/widgets/radio/__stories__/option-status.stories.jsx:25]: OptionStatus is only used by Choice and only once, so I took some of the logic for OptionStatus that was in Choice and moved it to OptionStatus.

File: packages/perseus/src/widgets/radio/base-radio.jsx [hunk 3]
```diff
@@ -47,6 +46,9 @@ const ChoicesType = PropTypes.arrayOf(
 const radioBorderColor = styleConstants.radioBorderColor;
 
 class BaseRadio extends React.Component<$FlowFixMe, $FlowFixMe> {
+    choiceRefs: Array<{current: ?HTMLDivElement}>;
+    state: $FlowFixMe;
```
Comment (handeyeco) [inline packages/perseus/src/widgets/radio/base-radio.jsx:50]: BaseRadio is on the docket to be refactored next. I'll fix this with that refactor.

File: packages/perseus/src/widgets/radio/base-radio.jsx [hunk 4]
```diff
@@ -340,9 +347,18 @@ class BaseRadio extends React.Component<$FlowFixMe, $FlowFixMe> {
     };
 
     focus: (number) => boolean = (i) => {
-        // $FlowFixMe[incompatible-use]
-        // $FlowFixMe[prop-missing]
-        ReactDOM.findDOMNode(this.refs["radio" + (i || 0)]).focus(); // eslint-disable-line react/no-string-refs
+        const ref = this.choiceRefs[i || 0];
```
Comment (handeyeco) [inline packages/perseus/src/widgets/radio/base-radio.jsx:350]: One of the big pieces when switching Choice to a functional component was that we could no longer stick a ref directly onto it. We still keep these refs in an array (now called choiceRefs), but we forward the refs down to the WB button; either:

```
BaseRadio -> Choice -> Clickable
or
BaseRadio -> ChoiceNoneAbove -> Choice -> Clickable
```

React's forwardRef combined with Flow's typing makes it kind of hard to read, but I like that it's more explicit now. Thanks Kevin for all your help with that.

File: packages/perseus/src/widgets/radio/base-radio.jsx [hunk 5]
```diff
@@ -340,9 +347,18 @@ class BaseRadio extends React.Component<$FlowFixMe, $FlowFixMe> {
...
+        // note(matthew): we know this is only getting passed
+        // to a WB Clickable button, so we force it to be of
+        // type HTMLButtonElement
+        const anyNode = (ReactDOM.findDOMNode(ref.current): any);
+        const castNode = (anyNode: ?HTMLButtonElement);
+
+        if (castNode) {
+            castNode.focus();
+        } else {
+            return false;
```
Comment (handeyeco) [inline packages/perseus/src/widgets/radio/base-radio.jsx:360]: Possible API change for Choice: before I think it would throw if the node didn't exist, now it returns false. I don't think this is an issue though.

File: packages/perseus/src/widgets/radio/base-radio.jsx [hunk 6]
```diff
@@ -363,60 +379,7 @@ class BaseRadio extends React.Component<$FlowFixMe, $FlowFixMe> {
...
-     * Find the choice at the given index (wrapped, if necessary), focus it,
-     * uncheck the old choice, and check the new choice (unless it's crossed
-     * out, in which case we only uncheck the old choice).
-     *
-     * This is very similar to standard up/down arrow behavior, except that we
-     * don't select crossed-out choices.
-     *
-     * Handles `goToPrevChoice` and `goToNextChoice` calls from our children.
-     */
-    goToChoice: (number) => void = (newChoiceIndex) => {
```
Comment (handeyeco) [inline packages/perseus/src/widgets/radio/base-radio.jsx:376]: A lot of the `goToXYZ` stuff seemed unused. I pulled it out of Choice and the removal rippled out a bit. I feel like I did my due diligence before removing it and the tests still passed. 🤷

File: packages/perseus/src/widgets/radio/choice-none-above.jsx [hunk 7]
```diff
@@ -1,54 +1,60 @@
...
+        ),
     };
 
-    static defaultProps: $FlowFixMe = {
-        showContent: true,
-    };
+    return <Choice {...choiceProps} ref={forwardedRef} />;
+}
 
-    focusInput: () => void = () => {
```
Comment (handeyeco) [inline packages/perseus/src/widgets/radio/choice-none-above.jsx:24]: I went ahead and refactored ChoiceNoneAbove at the same time as Choice since they're so closely related. Honestly they could probably be merged, but I didn't think about it until just now so I'll follow up later.

I couldn't find any uses of focusInput, so I pulled it out of these components.

File: packages/perseus/src/widgets/radio/choice-none-above.jsx [hunk 8]
```diff
@@ -1,54 +1,60 @@
...
-}
+type ExportProps = $Diff<
+    React.ElementConfig<typeof ChoiceNoneAbove>,
+    WithForwardRef,
+>;
 
-export default ChoiceNoneAbove;
+export default (React.forwardRef<ExportProps, HTMLButtonElement>(
+    (props, ref) => <ChoiceNoneAbove {...props} forwardedRef={ref} />,
+): React.AbstractComponent<ExportProps, HTMLButtonElement>);
```
Comment (handeyeco) [inline packages/perseus/src/widgets/radio/choice-none-above.jsx:60]: This is all just for forwarding a ref

File: packages/perseus/src/widgets/radio/choice.jsx [hunk 9]
```diff
@@ -9,192 +9,119 @@ import {Popover, PopoverContent} from "@khanacademy/wonder-blocks-popover";
...
-            }),
-            readOnly: PropTypes.bool,
-        }),
-        multipleSelect: PropTypes.bool,
-        checked: PropTypes.bool,
-        className: PropTypes.string,
-        rationale: PropTypes.node,
-        content: PropTypes.node,
-        correct: PropTypes.bool,
-        deselectEnabled: PropTypes.bool,
```
Comment (handeyeco) [inline packages/perseus/src/widgets/radio/choice.jsx:65]: I went through an pruned seemingly unused props like this one.

File: packages/perseus/src/widgets/radio/choice.jsx [hunk 10]
```diff
@@ -9,192 +9,119 @@ import {Popover, PopoverContent} from "@khanacademy/wonder-blocks-popover";
...
-        isInputActive: false,
-    };
-
-    UNSAFE_componentWillUpdate(nextProps: $FlowFixMe) {
-        if (this.state.isInputFocused && nextProps.disabled) {
-            this.setState({
-                isInputFocused: false,
-            });
-        }
+function ChoiceIconWrapper(props: ChoiceIconWrapperProps) {
```
Comment (handeyeco) [inline packages/perseus/src/widgets/radio/choice.jsx:41]: This was just a mini component exclusively being used by Choice. It was small enough that I didn't see the point of giving it its own file.

File: packages/perseus/src/widgets/radio/choice.jsx.fixture.js [hunk 11]
```diff
@@ -3,62 +3,44 @@ export const instances: $ReadOnlyArray<mixed> = [
...
         content: "Plain Content",
-        goToPrevChoice: () => {},
-        goToNextChoice: () => {},
     },
     {
         content: "Correct Content (Review)",
         correct: true,
         reviewMode: true,
-        goToPrevChoice: () => {},
-        goToNextChoice: () => {},
```
Comment (handeyeco) [inline packages/perseus/src/widgets/radio/choice.jsx.fixture.js:14]: Didn't see that this stuff was being used.

File: packages/perseus/src/widgets/radio/base-radio.jsx (hunk 4 above)
Comment (kevinbarabash) [reply packages/perseus/src/widgets/radio/base-radio.jsx:350]: So in either case the underlying ref will be the `Clickable`, is that right?

File: packages/perseus/src/widgets/radio/base-radio.jsx (hunk 6 above)
Comment (kevinbarabash) [reply packages/perseus/src/widgets/radio/base-radio.jsx:376]: I don't think we can actually rely on the tests passing as an indicator of whether or not it's being used.  If `goToChoice` is not being called by anything else then that's a good indicator.  The problem with relying solely on the tests to determine dead code is just b/c it isn't covered doesn't mean that it isn't used in production.

File: packages/perseus/src/widgets/radio/choice-none-above.jsx (hunk 7 above)
Comment (kevinbarabash) [reply packages/perseus/src/widgets/radio/choice-none-above.jsx:24]: There's a lot of stuff going on in this PR.  Merging can always happen in a future PR.

File: packages/perseus/src/widgets/radio/choice-none-above.jsx (hunk 8 above)
Comment (kevinbarabash) [reply packages/perseus/src/widgets/radio/choice-none-above.jsx:60]: Flow is just the worst when it comes to export types. 😞

File: packages/perseus/src/widgets/radio/base-radio.jsx (hunk 6 above)
Comment (kevinbarabash) [reply packages/perseus/src/widgets/radio/base-radio.jsx:376]: I had a look at the code and it looks like `goToChoice` is indeed not being used.

File: packages/perseus/src/widgets/radio/base-radio.jsx (hunk 5 above)
Comment (Myranae) [reply packages/perseus/src/widgets/radio/base-radio.jsx:360]: Question (non-blocking): So there isn't a time when an error would be useful from this? Does the false returned here ever get used for anything? If neither the error nor the false return are used, is an else section necessary?

File: packages/perseus/src/widgets/radio/base-radio.jsx (hunk 4 above)
Comment (Myranae) [reply packages/perseus/src/widgets/radio/base-radio.jsx:350]: Would love to pair sometime to learn more about refs and how they are affected by switching from class to functional. :)

File: packages/perseus/src/widgets/radio/choice-none-above.jsx (hunk 7 above)
Comment (Myranae) [reply packages/perseus/src/widgets/radio/choice-none-above.jsx:24]: Since they are so closely related, I also think merging in the future would make sense! Although, would that affect places where one is used without the other? Or are they always used together?

**Review by Myranae — COMMENTED** [review]
Wow, so many changes! I'm impressed you were able to find so many things that were not being used. Really interesting PR to review. I added some learning comments just with questions I am curious about, plus a couple small things that aren't very important but came to mind. There's a lot happening I don't quite understand, so I'll leave the approval to someone who understands it all :D

File: packages/perseus/src/widgets/__tests__/radio/choice_test.jsx [hunk 12]
```diff
@@ -0,0 +1,201 @@
+// @flow
```
Comment (kevinbarabash) [inline packages/perseus/src/widgets/__tests__/radio/choice_test.jsx:1]: I'm surprised that there weren't any tests for this before.  **praise:** thanks for adding these. 💚

File: packages/perseus/src/widgets/__tests__/radio/choice_test.jsx [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+
+    const overwrittenOptions = {...defaultOptions, ...options};
+
+    return render(<Choice {...overwrittenOptions} />);
+}
+
+describe("choice", () => {
+    it("renders choice content", () => {
+        // Arrange / Act
+        renderChoice();
```
[the remark this answers is not in the document]
Comment (kevinbarabash) [reply packages/perseus/src/widgets/__tests__/radio/choice_test.jsx:36]: I'm kind of surprised that Flow isn't complaining about no argument being passed to this.

File: packages/perseus/src/widgets/radio/base-radio.jsx (hunk 4 above)
Comment (handeyeco) [reply packages/perseus/src/widgets/radio/base-radio.jsx:350]: @kevinbarabash right, I don't see how there's any chance for it to not be a Clickable. Just sometimes it's further down the line than others.

@Myranae any time!

File: packages/perseus/src/widgets/radio/base-radio.jsx (hunk 5 above)
Comment (handeyeco) [reply packages/perseus/src/widgets/radio/base-radio.jsx:360]: @Myranae The more I think about it, the less I feel there's ever going to be a chance for this to fail/error. It should always end up being a Clickable which should have some kind of focusable element.

File: packages/perseus/src/widgets/__tests__/radio/choice_test.jsx [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+
+        // Assert
+        expect(screen.getByText("This is a good rationale")).toBeVisible();
+    });
+
+    it("renders choice button", () => {
+        // Arrange / Act
+        renderChoice();
+
+        const button = screen.getByRole("checkbox", {
```
[the remark this answers is not in the document]
Comment (kevinbarabash) [reply packages/perseus/src/widgets/__tests__/radio/choice_test.jsx:54]: Test code tends to be a bit more repetitive.  This makes it easier to understand what a test is doing by looking just at the `it()`.

File: packages/perseus/src/widgets/__tests__/radio/choice_test.jsx [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+        previouslyAnswered: false,
+        ref: {current: null},
+    };
+
+    const overwrittenOptions = {...defaultOptions, ...options};
+
+    return render(<Choice {...overwrittenOptions} />);
+}
+
+describe("choice", () => {
```
Comment (kevinbarabash) [inline packages/perseus/src/widgets/__tests__/radio/choice_test.jsx:33]: **question:** what kind of coverage is this test suite providing for choice.jsx?

File: packages/perseus/src/widgets/__tests__/radio/choice_test.jsx [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+            reviewMode: true,
+        });
+
+        // Assert
+        expect(screen.getByText("Incorrect (selected)")).toBeVisible();
+    });
+
+    it("can be checked", () => {
+        // Arrange
+        let checked = false;
```
[the remark this answers is not in the document]
Comment (kevinbarabash) [reply packages/perseus/src/widgets/__tests__/radio/choice_test.jsx:164]: `let` is necessary here because `checked` is updated in the `onChange` handler below.

File: packages/perseus/src/widgets/radio/__stories__/choice-none-above.stories.jsx [hunk 17]
```diff
@@ -17,10 +14,27 @@ type ChoiceNoneAboveStory = {|
...
+    disabled: false,
+    pos: 0,
+    reviewMode: false,
+    showRationale: false,
+    showCorrectness: false,
+    multipleSelect: false,
+    crossedOut: false,
+    previouslyAnswered: false,
+    apiOptions: {},
+    onChange: () => {},
```
Comment (kevinbarabash) [inline packages/perseus/src/widgets/radio/__stories__/choice-none-above.stories.jsx:34]: **suggest:** In some of the other stories we use the `actions` from `@storybook/addon-actions` to log when callbacks like this are called.  Can you do the same for `onChange`?

File: packages/perseus/src/widgets/__tests__/radio/choice_test.jsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (handeyeco) [reply packages/perseus/src/widgets/__tests__/radio/choice_test.jsx:36]: Yeah, that's a good question. Basically I'm defining default props to pass to Choice in the renderChoice function and giving consumers the option to overwrite the default props.

So take this example:

``` JS
function overwrite(options) {
  const base = {
    hello: "world"
  }
  return {...base, ...options}
}
console.log(overwrite())
// { hello: "world" }
console.log(overwrite(undefined)) // this is the same as the above example
// { hello: "world" }
console.log(overwrite({hello: "universe"}))
// { hello: "universe" }
console.log(overwrite({bye: "see ya!"}))
// { hello: "universe", bye: "see ya!" }
```

I originally had `options = {}` but Flow got fussy and it seemed more trouble than it was worth. So it's a way of generating an object with defaults that are easily overwritable.

File: packages/perseus/src/widgets/radio/__stories__/choice.stories.jsx [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+    correct: boolean,
+    disabled: boolean,
+    pos: number,
+    reviewMode: boolean,
+    showRationale: boolean,
+    showCorrectness: boolean,
+    multipleSelect: boolean,
+    crossedOut: boolean,
+    previouslyAnswered: boolean,
+|};
```
Comment (kevinbarabash) [inline packages/perseus/src/widgets/radio/__stories__/choice.stories.jsx:19]: **suggestion:** I think it would be good to include `onChange: () => {},` here as well.

File: packages/perseus/src/widgets/__tests__/radio/choice_test.jsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (handeyeco) [reply packages/perseus/src/widgets/__tests__/radio/choice_test.jsx:54]: Getting the button has to happen after renderChoice because there's no button until you call renderChoice. Or maybe I'm misunderstanding?

File: packages/perseus/src/widgets/radio/__stories__/choice.stories.jsx [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+    previouslyAnswered: false,
+};
+
+export default ({
+    title: "Perseus/Widgets/Radio/Choice",
+    args: defaultProps,
+}: Story);
+
+export const Interactive = (args: StoryArgs): React.Node => {
+    return <Choice {...args} />;
```
[the remark this answers is not in the document]
Comment (kevinbarabash) [reply packages/perseus/src/widgets/radio/__stories__/choice.stories.jsx:47]: **question:** why aren't we combining `defaultProps` with `args` like we are in the other stories?

File: packages/perseus/src/widgets/radio/base-radio.jsx [hunk 20]
```diff
@@ -340,9 +347,18 @@ class BaseRadio extends React.Component<$FlowFixMe, $FlowFixMe> {
...
     focus: (number) => boolean = (i) => {
-        // $FlowFixMe[incompatible-use]
-        // $FlowFixMe[prop-missing]
-        ReactDOM.findDOMNode(this.refs["radio" + (i || 0)]).focus(); // eslint-disable-line react/no-string-refs
+        const ref = this.choiceRefs[i || 0];
+        // note(matthew): we know this is only getting passed
+        // to a WB Clickable button, so we force it to be of
+        // type HTMLButtonElement
+        const anyNode = (ReactDOM.findDOMNode(ref.current): any);
+        const castNode = (anyNode: ?HTMLButtonElement);
```
Comment (kevinbarabash) [inline packages/perseus/src/widgets/radio/base-radio.jsx:355]: **style nit (non-blocking):** maybe `buttonNode`?

File: packages/perseus/src/widgets/__tests__/radio/choice_test.jsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (handeyeco) [reply packages/perseus/src/widgets/__tests__/radio/choice_test.jsx:164]: It's because it changes:

``` JS
        let checked = false; // this assigns checked to false
        renderChoice({
            checked,
            onChange: (next) => {
                checked = next.checked; // this reassigns checked to next.checked
            },
        });
```

Semantically `const` suggests something that doesn't change and `let` suggests something that does change. Or more accurately, `const` allows for only one assignment and `let` allows reassignment.

File: packages/perseus/src/widgets/radio/choice.jsx [hunk 21]
```diff
@@ -9,192 +9,119 @@ import {Popover, PopoverContent} from "@khanacademy/wonder-blocks-popover";
...
+}
 
-    renderOptionStatus: () => React.Node = () => {
-        const {correct, checked, crossedOut, reviewMode, previouslyAnswered} =
-            this.props;
+export type ChoiceProps = {|
+    // TODO(kevinb) use Options.propTypes from perseus-api.jsx
+    // This change will also require make sure that item-renderer.jsx and
+    // server-item-renderer.jsx have appropriate defaults for apiOptions
+    // because many of the properties on Options.propTypes are required.
```
Comment (kevinbarabash) [inline packages/perseus/src/widgets/radio/choice.jsx:59]: This comment is stale and can be removed now that we're using Flow and the `APIOptions` type.

File: packages/perseus/src/widgets/radio/__stories__/choice.stories.jsx (hunk 19 above)
[the remark this answers is not in the document]
Comment (handeyeco) [reply packages/perseus/src/widgets/radio/__stories__/choice.stories.jsx:47]: No, it's just what you need to do to pass an object of values as props to a React component: [Spread Attributes](https://reactjs.org/docs/jsx-in-depth.html#spread-attributes). I don't know what would happen if I changed it to `<Choice {args} />` honestly, I just know that we're supposed to use the `<Choice {...args} />` syntax. Usually you'll see it called props (ie `<Choice {...props} />`, but Storybook calls them args so I called them args. 🤷

File: packages/perseus/src/widgets/radio/base-radio.jsx [hunk 22]
```diff
@@ -468,8 +431,9 @@ class BaseRadio extends React.Component<$FlowFixMe, $FlowFixMe> {
                 <ul className={className} style={{listStyle: "none"}}>
                     {this.props.choices.map(function (choice, i) {
                         let Element = Choice;
+                        const ref = React.createRef();
+                        this.choiceRefs[i] = ref;
                         const elementProps = {
```
Comment (kevinbarabash) [inline packages/perseus/src/widgets/radio/base-radio.jsx:436]: **commentary:** I'm amazed that Flow is okay with this given that the `Choice` and `ChoiceAndNone` have slightly different prop types.

File: packages/perseus/src/widgets/radio/choice-none-above.jsx [hunk 23]
```diff
@@ -1,54 +1,60 @@
...
+                key="noneOfTheAboveRenderer"
+                content={i18n._("None of the above")}
+            />
+        ),
     };
 
-    static defaultProps: $FlowFixMe = {
-        showContent: true,
-    };
+    return <Choice {...choiceProps} ref={forwardedRef} />;
```
Comment (kevinbarabash) [inline packages/perseus/src/widgets/radio/choice-none-above.jsx:46]: **praise:** thanks for adding ref forwarding here as well.

**Review by kevinbarabash — APPROVED** [review]
LGTM.  Thank you for updating these components. 🎉 

There is a lot going on in this PR though.  In order to make things easier on reviewers please try to break up PR into smaller change sets.  In this case, choice_test.jsx and choice.stories.jsx could've been the first PR.  This would've increased confidence that the rest of the refactoring in this PR wasn't affecting the behaviour of the components.

File: packages/perseus/src/widgets/radio/__stories__/choice.stories.jsx (hunk 19 above)
[the remark this answers is not in the document]
Comment (handeyeco) [reply packages/perseus/src/widgets/radio/__stories__/choice.stories.jsx:47]: @kevinbarabash the Interactive story is the only one I want to change with the controls; the other stories I want to be fixed to show off specific states so I take the defaultProps and overwrite them directly.

defaultProps is being passed to the Storybook export which is overwritten by controls and passed to the Interactive story as args.

``` JS
const defaultProps = {
    checked: false,
    // ...etc...
};

export default ({
    title: "Perseus/Widgets/Radio/Choice",
    args: defaultProps, // <= passed to SB as args
}: Story);

export const Interactive = (args: StoryArgs) => { // <= received here after getting overwritten
    return <Choice {...args} />;
};
```

File: packages/perseus/src/widgets/radio/choice.jsx [hunk 24]
```diff
@@ -203,218 +130,240 @@ class Choice extends React.Component<$FlowFixMe, State> {
...
-        // content library.
-        const showDimmed =
-            (!sat && !reviewMode && this.props.apiOptions.readOnly) ||
-            this.props.crossedOut;
-
-        return (
+    function sendChange(newValues: {|
+        checked?: boolean,
+        crossedOut?: boolean,
+    |}) {
```
[the remark this answers is not in the document]
Comment (handeyeco) [reply packages/perseus/src/widgets/radio/choice.jsx:136]: Any particular reason? I know they're subtly different and in-vogue, but I consider arrow functions more difficult to parse (especially with Flow). I don't mind changing it, just curious.

File: packages/perseus/src/widgets/radio/base-radio.jsx (hunk 22 above)
Comment (handeyeco) [reply packages/perseus/src/widgets/radio/base-radio.jsx:436]: Further down it changes the props based on which component it's going to use:

```
if (choice.isNoneOfTheAbove) {
    Element = ChoiceNoneAbove;
    _.extend(elementProps, {
        showContent: choice.revealNoneOfTheAbove,
    });
}
```

Not sure if that clears anything up.

File: packages/perseus/src/widgets/__tests__/radio/choice_test.jsx [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+        expect(button.getAttribute("aria-disabled")).toBe("false");
+    });
+
+    it("shows correct when in review mode", () => {
+        // Arrange / Act
+        renderChoice({
+            checked: false,
+            correct: true,
+            showCorrectness: true,
+            reviewMode: true,
```
[the remark this answers is not in the document]
Comment (Myranae) [reply packages/perseus/src/widgets/__tests__/radio/choice_test.jsx:116]: Ooo, awesome! I will definitely play around with the story for this once it is released. Thank you!

File: packages/perseus/src/widgets/__tests__/radio/choice_test.jsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (Myranae) [reply packages/perseus/src/widgets/__tests__/radio/choice_test.jsx:164]: Oh wow. I think some of my learning on this was a little misled at some point. Thanks for clarifying. I thought const variables just couldn't change their type not that they couldn't be reassigned at all. Thank you both!

File: packages/perseus/src/widgets/__tests__/radio/choice_test.jsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (Myranae) [reply packages/perseus/src/widgets/__tests__/radio/choice_test.jsx:36]: This seems like a relaxed javascript thing perhaps? Usually you would need to define it as an optional parameter, right?

[the remark this answers is not in the document]
Comment (kevinbarabash) [reply packages/perseus/src/widgets/__tests__/radio/choice_test.jsx:36]: I forgot that Flow infers function params based on calls to that function as opposed to the function body.  It's such a weird behavior to me.

File: packages/perseus/src/widgets/__tests__/radio/choice_test.jsx (hunk 25 above)
[the remark this answers is not in the document]
Comment (kevinbarabash) [reply packages/perseus/src/widgets/__tests__/radio/choice_test.jsx:116]: @handeyeco it seems like the center line of the text isn't lined up with the center line of the icons.  Is this a known issue?

File: packages/perseus/src/widgets/radio/choice.jsx (hunk 24 above)
[the remark this answers is not in the document]
Comment (kevinbarabash) [reply packages/perseus/src/widgets/radio/choice.jsx:132]: Using arrow functions avoids issues involving `this`.  `this` isn't used here so it isn't necessary here, but it would be more consistent.  I get what you're saying about arrow functions being hard to parse when types are involved.  I updated my initial comment to be non-blocking so don't feel that you need to make this change.

File: packages/perseus/src/widgets/radio/base-radio.jsx (hunk 22 above)
Comment (kevinbarabash) [reply packages/perseus/src/widgets/radio/base-radio.jsx:436]: Right, but after that `elementProps.showContent` is `undefined | typeof choice.revealNoneOfTheAbove`.

File: packages/perseus/src/widgets/radio/choice.jsx [hunk 26]
```diff
@@ -9,192 +9,115 @@ import {Popover, PopoverContent} from "@khanacademy/wonder-blocks-popover";
...
-                multipleSelect={this.props.multipleSelect}
-                showCorrectness={this.props.showCorrectness}
-                reviewMode={this.props.reviewMode}
-                product={this.props.apiOptions.satStyling ? "sat" : "library"}
-                primaryProductColor={primaryProductColor}
-                previouslyAnswered={this.props.previouslyAnswered}
-            />
-        );
-    };
+    }, [disabled, isInputFocused, setIsInputFocused]);
```
Comment (kevinbarabash) [inline packages/perseus/src/widgets/radio/choice.jsx:120]: Thanks for adding the dep list.  We have a lint rule for this in webapp that we should add to Perseus.  I can add that in a separate PR.

File: packages/perseus/src/widgets/__tests__/radio/choice_test.jsx (hunk 25 above)
[the remark this answers is not in the document]
Comment (handeyeco) [reply packages/perseus/src/widgets/__tests__/radio/choice_test.jsx:116]: I checked and it's not a regression with this PR. I'll ping design and see if I need to file a new ticket.

**Comment by handeyeco** [comment]
Unless I missed it, I don't see any other asks so I'm going to merge this before the sprint ends.
